Formula/mesos.rb

```diff
@@ -53,6 +53,16 @@ class Mesos < Formula
   def install
     ENV.java_cache
 
+    # Disable optimizing as libc++ does not play well with optimized clang
+    # builds (see https://llvm.org/bugs/show_bug.cgi?id=28469 and
+    # https://issues.apache.org/jira/browse/MESOS-5745).
+    #
+    # NOTE: We cannot use `--disable-optimize` since we also pass e.g.,
+    # CXXFLAGS via environment variables. Since compiler flags are passed via
+    # environment variables the Mesos build will silently ignore flags like
+    # `--[disable|enable]-optimize`.
+    ENV.O0
```

Comment:
I tested this with the LLVM upstream reproducer. The test case linked there fails for any of `-Os`, `-O1`, or `-O2`, so I suspect the optimization pass exploiting the UB we are hitting is active in each of these optimization levels. The reproducer passes with `-O0`.